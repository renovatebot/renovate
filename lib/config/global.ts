import type { RenovateConfig, RepoGlobalConfig } from './types';

export class GlobalConfig {
  // TODO: once global config work is complete, add a test to make sure this list includes all options with globalOnly=true (#9603)
  private static readonly OPTIONS: (keyof RepoGlobalConfig)[] = [
    'allowCustomCrateRegistries',
    'allowedPostUpgradeCommands',
    'allowPlugins',
    'allowPostUpgradeCommandTemplating',
    'allowScripts',
    'binarySource',
    'cacheDir',
    'cacheHardTtlMinutes',
    'containerbaseDir',
    'customEnvVariables',
    'dockerChildPrefix',
    'dockerSidecarImage',
    'dockerUser',
    'dryRun',
    'exposeAllEnv',
    'executionTimeout',
    'githubTokenWarn',
    'localDir',
    'migratePresets',
    'privateKey',
    'privateKeyOld',
    'gitTimeout',
    'platform',
    'endpoint',
  ];

  private static config: RepoGlobalConfig = {};

  static get(): RepoGlobalConfig;
  static get<Key extends keyof RepoGlobalConfig>(
    key?: Key
  ): RepoGlobalConfig[Key];
  static get<Key extends keyof RepoGlobalConfig>(
    key?: Key
  ): RepoGlobalConfig | RepoGlobalConfig[Key] {
    return key ? GlobalConfig.config[key] : GlobalConfig.config;
  }

  static set(config: RenovateConfig | RepoGlobalConfig): RenovateConfig {
    GlobalConfig.reset();

    const result = { ...config };
    for (const option of GlobalConfig.OPTIONS) {
      GlobalConfig.config[option] = config[option] as never;
      delete result[option];
    }

    return result;
  }

  static reset(): void {
    GlobalConfig.config = {};
  }
}
