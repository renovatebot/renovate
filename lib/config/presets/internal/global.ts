import type { Preset } from '../types';

/* eslint sort-keys: ["error", "asc", {caseSensitive: false, natural: true}] */

export const presets: Record<string, Preset> = {
  safeEnv: {
    allowedEnv: ['GO*'],
    description:
      'Hopefully safe environment variables to allow users to configure.',
  },
};
