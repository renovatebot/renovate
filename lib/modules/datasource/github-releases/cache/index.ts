import type {
  CacheOptions,
  GithubCachedRelease,
  GithubGraphqlRelease,
} from '../../../../util/github/types';
import type { GithubHttp } from '../../../../util/http/github';
import { AbstractGithubDatasourceCache } from './cache-base';

export const query = `
query ($owner: String!, $name: String!, $cursor: String, $count: Int!) {
  repository(owner: $owner, name: $name) {
    payload: releases(
      first: $count
      after: $cursor
      orderBy: {field: CREATED_AT, direction: DESC}
    ) {
      nodes {
        version: tagName
        releaseTimestamp: publishedAt
        isDraft
        isPrerelease
        url
        id: databaseId
        name
        description
      }
      pageInfo {
        hasNextPage
        endCursor
      }
    }
  }
}
`;

export class CacheableGithubReleases extends AbstractGithubDatasourceCache<
  GithubCachedRelease,
  GithubGraphqlRelease
> {
  cacheNs = 'github-datasource-graphql-releases';
  graphqlQuery = query;

  constructor(http: GithubHttp, opts: CacheOptions = {}) {
    super(http, opts);
  }

  coerceFetched(item: GithubGraphqlRelease): GithubCachedRelease | null {
    const {
      version,
      releaseTimestamp,
      isDraft,
      isPrerelease,
      url,
      id,
      name,
      description,
    } = item;

    if (isDraft) {
      return null;
    }

    const result: GithubCachedRelease = {
      version,
      releaseTimestamp,
      url,
      id,
      name,
      description,
    };

    if (isPrerelease) {
      result.isStable = false;
    }

    return result;
  }
}
