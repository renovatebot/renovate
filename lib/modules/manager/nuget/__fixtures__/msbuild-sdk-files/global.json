{
  "sdk": {
    "version": "5.0.302",
    "rollForward": "latestMajor"
  },
  "msbuild-sdks": {
    "YoloDev.Sdk": "0.2.0"
  }
}
