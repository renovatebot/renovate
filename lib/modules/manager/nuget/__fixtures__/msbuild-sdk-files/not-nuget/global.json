{
  "type": "not a nuget global.json"
}
