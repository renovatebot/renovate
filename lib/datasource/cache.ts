import { logger } from '../logger';
import * as globalCache from '../util/cache/global';

/**
 * Cache callback result which has to be returned by the `CacheCallback` function.
 */
export interface CacheResult<TResult = unknown> {
  /**
   * The data which should be added to the cache
   */
  data: TResult;
  /**
   * `data` can only be cached if this is not `true`
   */
  isPrivate?: boolean;
}

/**
 * Simple helper type for defining the `CacheCallback` function return type
 */
export type CachePromise<TResult = unknown> = Promise<CacheResult<TResult>>;

/**
 * The callback function which is called on cache miss.
 */
export type CacheCallback<TArg, TResult = unknown> = (
  lookup: TArg
) => CachePromise<TResult>;

export type CacheConfig<TArg, TResult> = {
  /**
   * Datasource id
   */
  id: string;
  /**
   * Cache key
   */
  lookup: TArg;
  /**
   * Callback to use on cache miss to load result
   */
  cb: CacheCallback<TArg, TResult>;
  /**
   * Time to cache result in minutes
   */
  minutes?: number;
};

/**
 * Loads result from cache or from passed callback on cache miss.
 * @param param0 Cache config args
 */
export async function cacheAble<TArg, TResult = unknown>({
  id,
  lookup,
  cb,
  minutes = 60,
}: CacheConfig<TArg, TResult>): Promise<TResult> {
  const cacheNamespace = `datasource-${id}`;
  const cacheKey = JSON.stringify(lookup);
  const cachedResult = await globalCache.get<TResult>(cacheNamespace, cacheKey);
  // istanbul ignore if
  if (cachedResult) {
    logger.trace({ id, lookup }, 'datasource cachedResult');
    return cachedResult;
  }
  const { data, isPrivate } = await cb(lookup);
  // istanbul ignore if
  if (isPrivate) {
    logger.trace({ id, lookup }, 'Skipping datasource cache for private data');
  } else {
    await globalCache.set(cacheNamespace, cacheKey, data, minutes);
  }
  return data;
}
