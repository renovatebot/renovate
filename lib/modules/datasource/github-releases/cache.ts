import { DateTime, DurationLike } from 'luxon';
import { logger } from '../../../logger';
import * as packageCache from '../../../util/cache/package';
import type {
  GithubGraphqlResponse,
  GithubHttp,
} from '../../../util/http/github';
import type { GetReleasesConfig, Release } from '../types';
import { getApiBaseUrl } from './common';

const softReset: DurationLike = { minutes: 30 };
const hardReset: DurationLike = { days: 7 };
const stabilityPeriod: DurationLike = { months: 1 };

const query = `
query(
  $owner: String!,
  $name: String!,
  $cursor: String,
  $count: Int!
) {
  repository(owner: $owner, name: $name) {
    releases(after: $cursor, first: $count, orderBy: {field: CREATED_AT, direction: DESC}) {
      nodes {
        version: tagName
        gitRef: tagName
        releaseTimestamp: publishedAt
        isDraft
        isPrerelease
        updatedAt
      }

      pageInfo {
        hasNextPage
        endCursor
      }
    }
  }
}
`;

interface QueryParams {
  owner: string;
  name: string;
  cursor: string | null;
  count: number;
}

interface QueryResponse {
  repository: {
    releases: {
      nodes: FetchedItem[];
      pageInfo: {
        hasNextPage: boolean;
        endCursor: string;
      };
    };
  };
}

interface ItemBase {
  version: string;
  releaseTimestamp: string;
}

interface FetchedItem extends ItemBase {
  isDraft: boolean;
  isPrerelease: boolean;
  updatedAt: string;
}

interface CachedItem extends ItemBase {
  gitRef: string;
  isStable?: boolean;
}

interface GithubReleasesCache {
  releases: Record<string, CachedItem>;
  releasesUpdatedAt: Record<string, string>;
  createdAt: string;
  updatedAt: string;
}

function coerceFetchedItem(item: FetchedItem): CachedItem {
  const { version, releaseTimestamp, isDraft, isPrerelease } = item;
  const gitRef = version;
  const result: CachedItem = { version, gitRef, releaseTimestamp };
  if (isPrerelease || isDraft) {
    result.isStable = false;
  }
  return result;
}

function createCache(now: DateTime): GithubReleasesCache {
  return {
    releases: {},
    releasesUpdatedAt: {},
    createdAt: now.toISO(),
    updatedAt: now.minus(softReset).toISO(),
  };
}

function isExpired(
  now: DateTime,
  date: string,
  duration: DurationLike
): boolean {
  const then = DateTime.fromISO(date);
  return now >= then.plus(duration);
}

export async function getCacheableReleases(
  http: GithubHttp,
  config: GetReleasesConfig
): Promise<Release[]> {
  const now = DateTime.now();
  let cache: GithubReleasesCache = createCache(now);
  const { packageName, registryUrl } = config;

  const baseUrl = getApiBaseUrl(registryUrl).replace('/v3/', '/');
  const [owner, name] = packageName.split('/');
  if (owner && name) {
    const cacheNs = `datasource-github-releases-graphql`;
    const cacheKey = `${baseUrl}:${owner}/${name}`;
    const cachedRes = await packageCache.get<GithubReleasesCache>(
      cacheNs,
      cacheKey
    );

    let isCacheUpdated = false;

    if (cachedRes && !isExpired(now, cachedRes.createdAt, hardReset)) {
      cache = cachedRes;
    } else {
      isCacheUpdated = true;
    }

    if (isExpired(now, cache.updatedAt, softReset)) {
      const variables: QueryParams = {
        owner,
        name,
        cursor: null,
        count: 100,
      };

      const checkedItems = new Set<string>();

      try {
        let pagesAllowed = 10;
        let isIterating = true;
        while (pagesAllowed > 0 && isIterating) {
          const graphqlRes = await http.postJson<
            GithubGraphqlResponse<QueryResponse>
          >('/graphql', {
            baseUrl,
            body: { query, variables },
          });
          pagesAllowed -= 1;

          const data = graphqlRes.body.data;
          if (data) {
            const {
              nodes: fetchedItems,
              pageInfo: { hasNextPage, endCursor },
            } = data.repository.releases;

            if (!hasNextPage) {
              isIterating = false;
            } else {
              variables.cursor = endCursor;
            }

            for (const item of fetchedItems) {
              const { version, releaseTimestamp, updatedAt } = item;
              checkedItems.add(version);

              const storedItem = cache.releases[version];
              const storedItemUpdatedAt = cache.releasesUpdatedAt[version];
              if (!storedItem || storedItemUpdatedAt !== updatedAt) {
                cache.releases[version] = coerceFetchedItem(item);
                isCacheUpdated = true;
              } else if (isExpired(now, releaseTimestamp, stabilityPeriod)) {
                isIterating = false;
                break;
              }
            }
          }
        }
      } catch (err) {
        logger.debug(
          { err },
          `GitHub releases: error fetching cacheable GraphQL data`
        );
      }

      for (const [version, item] of Object.entries(cache.releases)) {
        if (
          !isExpired(now, item.releaseTimestamp, stabilityPeriod) &&
          !checkedItems.has(version)
        ) {
          delete cache.releases[version];
          isCacheUpdated = true;
        }
      }

      if (isCacheUpdated) {
        const expiry = DateTime.fromISO(cache.createdAt).plus(hardReset);
        const { minutes: ttlMinutes } = expiry
          .diff(now, ['minutes'])
          .toObject();
        if (ttlMinutes && ttlMinutes > 0) {
          cache.updatedAt = now.toISO();
          await packageCache.set(cacheNs, cacheKey, cache, ttlMinutes);
        }
      }
    }
  }

  const result = Object.values(cache.releases);
  return result;
}
