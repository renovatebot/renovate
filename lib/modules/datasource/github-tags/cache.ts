import type {
  CacheOptions,
  GithubCachedTag,
  GithubGraphqlTag,
} from '../../../util/github/types';
import type { GithubHttp } from '../../../util/http/github';
import { AbstractGithubDatasourceCache } from '../github-releases/cache/cache-base';

const query = `
query ($owner: String!, $name: String!, $cursor: String, $count: Int!) {
  repository(owner: $owner, name: $name) {
    payload: refs(
      first: $count
      after: $cursor
      orderBy: {field: TAG_COMMIT_DATE, direction: DESC}
      refPrefix: "refs/tags/"
    ) {
      nodes {
        version: name
        target {
          type: __typename
          ... on Commit {
            hash: oid
            releaseTimestamp: committedDate
          }
          ... on Tag {
            target {
              ... on Commit {
                hash: oid
              }
            }
            tagger {
              releaseTimestamp: date
            }
          }
        }
      }
      pageInfo {
        hasNextPage
        endCursor
      }
    }
  }
}
`;

export class CacheableGithubTags extends AbstractGithubDatasourceCache<
  GithubCachedTag,
  GithubGraphqlTag
> {
  readonly cacheNs = 'github-datasource-graphql-tags-v2';
  readonly graphqlQuery = query;

  constructor(http: GithubHttp, opts: CacheOptions = {}) {
    super(http, opts);
  }

  coerceFetched(item: GithubGraphqlTag): GithubCachedTag | null {
    const { version, target } = item;
    if (target.type === 'Commit') {
      const { hash, releaseTimestamp } = target;
      return { version, hash, releaseTimestamp };
    } else if (target.type === 'Tag') {
      const { hash } = target.target;
      const { releaseTimestamp } = target.tagger;
      return { version, hash, releaseTimestamp };
    }
    return null;
  }
}
