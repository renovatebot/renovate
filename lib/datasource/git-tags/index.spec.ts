import fs from 'fs-extra';
import _simpleGit from 'simple-git/promise';
import * as globalCache from '../../util/cache/global';
import { getDigest, getReleases } from '.';

jest.mock('simple-git/promise');
const simpleGit: any = _simpleGit;

const lookupName = 'https://github.com/example/example.git';

const lsRemote1 = fs.readFileSync(
  'lib/datasource/git-refs/__fixtures__/ls-remote-1.txt',
  'utf8'
);

describe('datasource/git-tags', () => {
  beforeEach(() => globalCache.rmAll());
  describe('getReleases', () => {
    it('returns nil if response is wrong', async () => {
      simpleGit.mockReturnValue({
        listRemote() {
          return Promise.resolve(null);
        },
      });
      const versions = await getReleases({ lookupName });
      expect(versions).toEqual(null);
    });
    it('returns nil if remote call throws exception', async () => {
      simpleGit.mockReturnValue({
        listRemote() {
          throw new Error();
        },
      });
      const versions = await getReleases({ lookupName });
      expect(versions).toEqual(null);
    });
    it('returns versions filtered from tags', async () => {
      simpleGit.mockReturnValue({
        listRemote() {
          return Promise.resolve(lsRemote1);
        },
      });

      const versions = await getReleases({
        lookupName,
      });
      expect(versions).toMatchSnapshot();
    });
  });
  describe('getDigest()', () => {
    it('returns null if not found', async () => {
      simpleGit.mockReturnValue({
        listRemote() {
          return Promise.resolve(lsRemote1);
        },
      });
      const digest = await getDigest(
        { lookupName: 'a tag to look up' },
        'notfound'
      );
      expect(digest).toBeNull();
    });
    it('returns digest for tag', async () => {
      simpleGit.mockReturnValue({
        listRemote() {
          return Promise.resolve(lsRemote1);
        },
      });
      const digest = await getDigest(
        { lookupName: 'a tag to look up' },
        'v1.0.2'
      );
      expect(digest).toMatchSnapshot();
    });
    it('returns digest for HEAD', async () => {
      simpleGit.mockReturnValue({
        listRemote() {
          return Promise.resolve(lsRemote1);
        },
      });
      const digest = await getDigest(
        { lookupName: 'another tag to look up' },
        undefined
      );
      expect(digest).toMatchSnapshot();
    });
  });
});
