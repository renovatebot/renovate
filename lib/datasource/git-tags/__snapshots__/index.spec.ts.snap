// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`datasource/git-tags getDigest() returns digest for HEAD 1`] = `"a9920c014aebc28dc1b23e7efcc006d0455cc710"`;

exports[`datasource/git-tags getDigest() returns digest for tag 1`] = `"9cb93e0b236385a4e2efd089d7c6a458f5ff321f"`;

exports[`datasource/git-tags getReleases returns versions filtered from tags 1`] = `
Object {
  "releases": Array [
    Object {
      "gitRef": "v1.0.0",
      "newDigest": "0be9ffb1ae0c8d0846cd88b58dfef42d74674673",
      "version": "v1.0.0",
    },
    Object {
      "gitRef": "v1.0.1",
      "newDigest": "281fbfb58990ec98b237a923d67904c102bec34c",
      "version": "v1.0.1",
    },
    Object {
      "gitRef": "v1.0.2",
      "newDigest": "9cb93e0b236385a4e2efd089d7c6a458f5ff321f",
      "version": "v1.0.2",
    },
    Object {
      "gitRef": "v1.0.3",
      "newDigest": "8b0d0e0aec21ea059448ef0255387dbb82c61973",
      "version": "v1.0.3",
    },
    Object {
      "gitRef": "v1.0.4",
      "newDigest": "2b52829c7c1bd65b3501c450849c53b90b11fa0e",
      "version": "v1.0.4",
    },
    Object {
      "gitRef": "v1.0.5",
      "newDigest": "2d138c34e4c6939d0a8686943e851c6528aa04db",
      "version": "v1.0.5",
    },
  ],
  "sourceUrl": "https://github.com/example/example",
}
`;
